src/js/tracing/nativeframes.ts

```diff
@@ -0,0 +1,263 @@
+import { Span, Transaction } from "@sentry/tracing";
+import { Event, EventProcessor, Measurements } from "@sentry/types";
+import { logger, timestampInSeconds } from "@sentry/utils";
+
+import { NativeFramesResponse } from "../definitions";
+import { NATIVE } from "../wrapper";
+import { instrumentChildSpanFinish } from "./utils";
+
+interface FramesMeasurements extends Measurements {
+  frames_total: {
+    value: number;
+  };
+  frames_slow: {
+    value: number;
+  };
+  frames_frozen: {
+    value: number;
+  };
+}
+
+const MARGIN_OF_ERROR_SECONDS = 0.05;
+
+/**
+ * Instrumentation to add native slow/frozen frames measurements onto transactions.
+ */
+export class NativeFramesInstrumentation {
+  private _finishFrames: Map<
+    string,
+    { timestamp: number; nativeFrames: NativeFramesResponse | null }
+  > = new Map();
+  private _framesListeners: Map<string, () => void> = new Map();
+  private _lastSpanFinishFrames?: {
+    timestamp: number;
+    nativeFrames: NativeFramesResponse;
+  };
+
+  public constructor(
+    addGlobalEventProcessor: (e: EventProcessor) => void,
+    doesExist: () => boolean
+  ) {
+    logger.log(
+      "[ReactNativeTracing] Native frames instrumentation initialized."
+    );
+
+    addGlobalEventProcessor((event) => {
+      if (doesExist()) {
+        return this._processEvent(event);
+      }
+
+      return event;
+    });
+  }
+
+  /**
+   * To be called when a transaction is started
+   */
+  public onTransactionStart(transaction: Transaction): void {
+    void NATIVE.fetchNativeFrames().then((framesMetrics) => {
+      if (framesMetrics) {
+        transaction.setData("__startFrames", framesMetrics);
+      }
+    });
+
+    instrumentChildSpanFinish(transaction, this._onSpanFinish.bind(this));
+  }
+
+  /**
+   * To be called when a transaction is finished
+   */
+  public onTransactionFinish(transaction: Transaction): void {
+    void this._fetchFramesForTransaction(transaction);
+  }
+
+  /**
+   * Called on a span finish to fetch native frames to support transactions with trimEnd.
+   */
+  private _onSpanFinish(_: Span, endTimestamp?: number): void {
+    if (!endTimestamp) {
```

Comment:
If lack of an `endTimestamp` means that logic should be run, can we just change so that `_onSpanFinish` is only called if an end timestamp doesn't exist (move the logic outside of this func?)